src/util/async/Operation.hpp

```diff
@@ -0,0 +1,169 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2024, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+#include "util/async/Concepts.hpp"
+#include "util/async/Outcome.hpp"
+#include "util/async/context/impl/Cancellation.hpp"
+#include "util/async/context/impl/Timer.hpp"
+
+#include <condition_variable>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <optional>
+
+namespace util::async {
+namespace detail {
+
+template <typename OutcomeType>
+class BasicOperation {
+protected:
+    std::future<typename OutcomeType::DataType> future_;
+
+public:
+    using DataType = typename OutcomeType::DataType;
+
+    explicit BasicOperation(OutcomeType* outcome) : future_{outcome->getStdFuture()}
+    {
+    }
+
+    BasicOperation(BasicOperation&&) = default;
+    BasicOperation(BasicOperation const&) = delete;
+
+    [[nodiscard]] auto
+    get()
+    {
+        return future_.get();
+    }
+
+    void
+    wait()
+    {
+        future_.wait();
+    }
+};
+
+template <typename CtxType, typename OpType>
+struct BasicScheduledOperation {
+    struct State {
+        std::mutex m_;
+        std::condition_variable ready_;
+        std::optional<OpType> op_{std::nullopt};
+
+        void
+        emplace(auto&& op)
+        {
+            std::lock_guard lock{m_};
+            op_.emplace(std::forward<decltype(op)>(op));
+            ready_.notify_all();
+        }
+
+        [[nodiscard]] OpType&
+        get()
+        {
+            std::unique_lock lock{m_};
+            ready_.wait(lock, [this] { return op_.has_value(); });
+            return op_.value();
+        }
+    };
+
+    std::shared_ptr<State> state_ = std::make_shared<State>();
+    typename CtxType::Timer timer_;
+
+    BasicScheduledOperation(auto& executor, auto delay, auto&& fn)
+        : timer_(executor, delay, [state = state_, fn = std::forward<decltype(fn)>(fn)](auto ec) mutable {
+            state->emplace(fn(ec));
```

Comment:
Shall we check ec here , if canceled : boost::asio::error::operation_aborted, not to call fn at all?

Reply:
This is intentionally passed down to the erased execution context. Without it we would not be able to call user's callback with two args, second one being `bool isCancelled`.
It's handled by `isAborted = (ec == boost::asio::error::operation_aborted)` in `BasicExecutionContext.hpp`.